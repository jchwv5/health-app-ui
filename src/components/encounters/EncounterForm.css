.formPopup {
    position:absolute;
    top:30%;
    left:25%;    
    margin-top:-300px;
    width: 800px;
    height: 800px;
    background-color: white;
    opacity: 100%;
  }
  
  .formContainer {
    max-width: 100%;
    padding: 10px;
    
  }
  
  .form {
    display: flex;  }
  
  .column {
    flex: 33%;
  }
  
  .formHeader {
    text-align: center;
  }
  
  .btn {
    background-color: #04AA6D;
    color: white;
    padding: 16px 20px;
    border-radius: 20px;
    cursor: pointer;
    width: 20%;
    float: left;
    margin-bottom:10px;
    margin-left: 150px;
    opacity: 0.8;
  }
  
  .cancel {
    background-color: red;
    margin-left: 100px;
  }
  
  .btn:hover, .open-button:hover {
    opacity: 1;
  }
  
  .buttonWrapper {
    align-items: left;
    background-color: white;
  }
  
  .errorDiv {
    height: 40px;
    color: red;
    font-size: 14px;
    margin: 15px 2px 4px 4px;
  }
  
  .serverMsgDiv {
    height: 40px;
    font-size: 14px;
    margin: 0px 2px 4px 4px;
  }
  
  .successMsg {
    color: green;
  }
  
  .failureMsg {
    color: red;
  }
  
  .input {
    position: fixed;
  }
  