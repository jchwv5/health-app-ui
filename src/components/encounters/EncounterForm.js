/* eslint-disable no-unused-vars */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState } from 'react';
import './EncounterForm.css';
import FormItem from '../../utils/FormItem';
import HandleEncounterApiCall from './HandleEncounterApiCall';

/**
 * Modular form to be used for adding new encounters and updating existing encounters
 * @param {*} props - takes in parameters from onClick that calls the form to determine whether
 * form will be used for adding a new encounter or updating an existing one
 * @returns - form component
 */

const EncounterForm = (props) => {
  const { onClose } = props;
  const { encounterApiCall } = props;
  const { patientId } = props;
  const { encounterId } = props;
  const { formHeader } = props;
  const { buttonLabel } = props;
  const { successMsg } = props;
  const [encounterData, setEncounterData] = useState({});
  const [apiError, setApiError] = useState(false);
  const [encounterSuccess, setEncounterSuccess] = useState(false);
  const [encounterFailure, setEncounterFailure] = useState(false);
  const [encounterErrors, setEncounterErrors] = useState({
    visitCode: { dataIsValid: false, errorMessage: '' },
    provider: { dataIsValid: false, errorMessage: '' },
    billingCode: { dataIsValid: false, errorMessage: '' },
    icd10: { dataIsValid: false, errorMessage: '' },
    totalCost: { dataIsValid: false, errorMessage: '' },
    copay: { dataIsValid: false, errorMessage: '' },
    chiefComplaint: { dataIsValid: false, errorMessage: '' },
    pulse: { dataIsValid: false, errorMessage: '' },
    systolic: { dataIsValid: false, errorMessage: '' },
    diastolic: { dataIsValid: false, errorMessage: '' },
    date: { dataIsValid: false, errorMessage: '' }
  });
  /**
 * function to take user input and assign it to the value for that field
 * @param {*} e - input field that is receiving user input
 */
  const handleChange = (e) => {
    setEncounterData((prevValue) => ({ ...prevValue, [e.target.id]: e.target.value }));
  };

  return (
    <div className="formPopup" id="myForm">
      <form action="/action_page.php" className="formContainer">
        <h1 className="formHeader">{formHeader}</h1>
        <p>(fields with a * are required)</p>
        <div className="form">
          <div className="column">
            <FormItem
              className="input"
              type="text"
              id="notes"
              label="Notes"
              placeholder="Please enter any notes"
              onChange={handleChange}
              value={encounterData.notes}
            />
            <FormItem
              className="input"
              type="text"
              id="visitCode"
              label="Visit Code*"
              placeholder="ex. H1H 1H1"
              onChange={handleChange}
              value={encounterData.visitCode}
            />
            <div className="errorDiv">
              {encounterErrors.visitCode.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.visitCode.errorMessage}</p>
              )}
            </div>
            <FormItem
              className="input"
              type="text"
              id="provider"
              label="Provider*"
              placeholder="ex. Dr. Mitchell"
              onChange={handleChange}
              value={encounterData.provider}
            />
            <div className="errorDiv">
              {encounterErrors.provider.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.provider.errorMessage}</p>
              )}
            </div>
            <FormItem
              className="input"
              type="text"
              id="billingCode"
              label="Billing Code*"
              placeholder="ex. 123.456.789-00"
              onChange={handleChange}
              value={encounterData.billingCode}
            />
            <div className="errorDiv">
              {encounterErrors.billingCode.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.billingCode.errorMessage}</p>
              )}
            </div>
          </div>
          <div className="column">
            <FormItem
              className="input"
              type="text"
              id="icd10"
              label="ICD10*"
              placeholder="ex. A12"
              onChange={handleChange}
              value={encounterData.icd10}
            />
            <div className="errorDiv">
              {encounterErrors.icd10.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.icd10.errorMessage}</p>
              )}
            </div>
            <FormItem
              className="input"
              type="text"
              id="totalCost"
              label="Total Cost*"
              placeholder="ex. 1000"
              onChange={handleChange}
              value={encounterData.totalCost}
            />
            <div className="errorDiv">
              {encounterErrors.totalCost.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.totalCost.errorMessage}</p>
              )}
            </div>
            <FormItem
              className="input"
              type="text"
              id="copay"
              label="Copay*"
              placeholder="ex. 100"
              onChange={handleChange}
              value={encounterData.copay}
            />
            <div className="errorDiv">
              {encounterErrors.copay.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.copay.errorMessage}</p>
              )}
            </div>
            <FormItem
              className="input"
              type="text"
              id="chiefComplaint"
              label="Chief Complaint*"
              placeholder="ex. Broken Arm"
              onChange={handleChange}
              value={encounterData.chiefComplaint}
            />
            <div className="errorDiv">
              {encounterErrors.chiefComplaint.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.chiefComplaint.errorMessage}</p>
              )}
            </div>
          </div>
          <div className="column">
            <FormItem
              className="input"
              type="text"
              id="pulse"
              label="Pulse"
              placeholder="ex. 90"
              onChange={handleChange}
              value={encounterData.pulse}
            />
            <div className="errorDiv">
              {encounterErrors.pulse.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.pulse.errorMessage}</p>
              )}
            </div>
            <FormItem
              className="input"
              type="text"
              id="systolic"
              label="Systolic"
              placeholder="ex. 100"
              onChange={handleChange}
              value={encounterData.systolic}
            />
            <div className="errorDiv">
              {encounterErrors.systolic.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.systolic.errorMessage}</p>
              )}
            </div>
            <FormItem
              className="input"
              type="text"
              id="diastolic"
              label="Diastolic"
              placeholder="ex. 88"
              onChange={handleChange}
              value={encounterData.diastolic}
            />
            <div className="errorDiv">
              {encounterErrors.diastolic.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.diastolic.errorMessage}</p>
              )}
            </div>
            <FormItem
              className="input"
              type="text"
              id="date"
              label="Encounter Date*"
              placeholder="ex. 2020-01-01"
              onChange={handleChange}
              value={encounterData.date}
            />
            <div className="errorDiv">
              {encounterErrors.date.dataIsValid === false && (
              <p className="errorMsg">{encounterErrors.date.errorMessage}</p>
              )}
            </div>
          </div>
        </div>
      </form>
      <div className="buttonWrapper">
        <div className="serverMsgDiv">
          {encounterSuccess === true && (
          <p className="successMsg">{successMsg}</p>
          )}
          {encounterFailure === true && (
          <p className="failureMsg">Oops! Something went wrong!</p>
          )}
        </div>
        <button
          type="submit"
          className="btn"
          id="callButton"
          onClick={() => HandleEncounterApiCall(encounterData,
            patientId,
            encounterId,
            setEncounterSuccess,
            setEncounterFailure,
            encounterErrors,
            setEncounterErrors,
            setApiError,
            encounterApiCall)}
        >
          {buttonLabel}
        </button>
        <button type="button" className="btn cancel" onClick={onClose}>Close</button>
      </div>
    </div>

  );
};

export default EncounterForm;
