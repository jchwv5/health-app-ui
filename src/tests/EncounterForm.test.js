import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import ReactDom from 'react-dom';
import renderer from 'react-test-renderer';
import EncounterForm from '../components/encounters/EncounterForm';

it('renders successfully', () => {
  const div = document.createElement('div');
  ReactDom.render(<EncounterForm />, div);
  ReactDom.unmountComponentAtNode(div);
});

it('App component matches snapshot', () => {
  const component = renderer.create(<EncounterForm />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('form validation fails empty form', () => {
  const { getByLabelText, getByText } = render(<EncounterForm buttonLabel="Edit Encounter" />);
  fireEvent.change(getByLabelText('Visit Code*'), { target: { value: '' } });
  fireEvent.change(getByLabelText('Provider*'), { target: { value: '' } });
  fireEvent.change(getByLabelText('Billing Code*'), { target: { value: '' } });
  fireEvent.change(getByLabelText('ICD10*'), { target: { value: '' } });
  fireEvent.change(getByLabelText('Total Cost*'), { target: { value: '' } });
  fireEvent.change(getByLabelText('Copay*'), { target: { value: '' } });
  fireEvent.change(getByLabelText('Chief Complaint*'), { target: { value: '' } });
  fireEvent.change(getByLabelText('Encounter Date*'), { target: { value: '' } });
  fireEvent.click(getByText('Edit Encounter'));

  expect(getByText('Visit Code field must not be left empty',
    'Provider field must not be left empty',
    'Billing Code field must not be left empty',
    'ICD10 field must not be left empty',
    'Total Cost field must not be left empty',
    'Copay field must not be left empty',
    'Chief Complaint field must not be left empty',
    'Encounter Date field must not be left empty')).toBeInTheDocument();
});
