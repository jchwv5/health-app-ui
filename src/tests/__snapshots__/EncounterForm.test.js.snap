// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component matches snapshot 1`] = `
<div
  className="formPopup"
  id="myForm"
>
  <form
    action="/action_page.php"
    className="formContainer"
  >
    <h1
      className="formHeader"
    />
    <p>
      (fields with a * are required)
    </p>
    <div
      className="form"
    >
      <div
        className="column"
      >
        <div>
          <label
            className="label"
            htmlFor="notes"
          >
            Notes
            <div>
              <input
                className="input"
                id="notes"
                onChange={[Function]}
                placeholder="Please enter any notes"
                type="text"
              />
            </div>
          </label>
        </div>
        <div>
          <label
            className="label"
            htmlFor="visitCode"
          >
            Visit Code*
            <div>
              <input
                className="input"
                id="visitCode"
                onChange={[Function]}
                placeholder="ex. H1H 1H1"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
        <div>
          <label
            className="label"
            htmlFor="provider"
          >
            Provider*
            <div>
              <input
                className="input"
                id="provider"
                onChange={[Function]}
                placeholder="ex. Dr. Mitchell"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
        <div>
          <label
            className="label"
            htmlFor="billingCode"
          >
            Billing Code*
            <div>
              <input
                className="input"
                id="billingCode"
                onChange={[Function]}
                placeholder="ex. 123.456.789-00"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
      </div>
      <div
        className="column"
      >
        <div>
          <label
            className="label"
            htmlFor="icd10"
          >
            ICD10*
            <div>
              <input
                className="input"
                id="icd10"
                onChange={[Function]}
                placeholder="ex. A12"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
        <div>
          <label
            className="label"
            htmlFor="totalCost"
          >
            Total Cost*
            <div>
              <input
                className="input"
                id="totalCost"
                onChange={[Function]}
                placeholder="ex. 1000"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
        <div>
          <label
            className="label"
            htmlFor="copay"
          >
            Copay*
            <div>
              <input
                className="input"
                id="copay"
                onChange={[Function]}
                placeholder="ex. 100"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
        <div>
          <label
            className="label"
            htmlFor="chiefComplaint"
          >
            Chief Complaint*
            <div>
              <input
                className="input"
                id="chiefComplaint"
                onChange={[Function]}
                placeholder="ex. Broken Arm"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
      </div>
      <div
        className="column"
      >
        <div>
          <label
            className="label"
            htmlFor="pulse"
          >
            Pulse
            <div>
              <input
                className="input"
                id="pulse"
                onChange={[Function]}
                placeholder="ex. 90"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
        <div>
          <label
            className="label"
            htmlFor="systolic"
          >
            Systolic
            <div>
              <input
                className="input"
                id="systolic"
                onChange={[Function]}
                placeholder="ex. 100"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
        <div>
          <label
            className="label"
            htmlFor="diastolic"
          >
            Diastolic
            <div>
              <input
                className="input"
                id="diastolic"
                onChange={[Function]}
                placeholder="ex. 88"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
        <div>
          <label
            className="label"
            htmlFor="date"
          >
            Encounter Date*
            <div>
              <input
                className="input"
                id="date"
                onChange={[Function]}
                placeholder="ex. 2020-01-01"
                type="text"
              />
            </div>
          </label>
        </div>
        <div
          className="errorDiv"
        >
          <p
            className="errorMsg"
          >
            
          </p>
        </div>
      </div>
    </div>
  </form>
  <div
    className="buttonWrapper"
  >
    <div
      className="serverMsgDiv"
    />
    <button
      className="btn"
      id="callButton"
      onClick={[Function]}
      type="submit"
    />
    <button
      className="btn cancel"
      type="button"
    >
      Close
    </button>
  </div>
</div>
`;
